src/veranda/io/geotiff.py

```diff
@@ -341,135 +311,160 @@ def __init__(self, filename, mode='r', count=None, compression='LZW',
+        n_rows : int, optional
+            Number of rows to read (default is 1).
+        n_cols : int, optional
+            Number of columns to read (default is 1).
+        band : int or list of int, optional
+            Band numbers (starting with 1). If None, all bands will be read.
+        nodataval : tuple or list, optional
+            List of no data values for each band.
+            Default: -9999 for each band.
+        decoder : function, optional
+            Decoding function expecting a NumPy array as input.
+        decoder_kwargs : dict, optional
+            Keyword arguments for the decoder.
 
         Returns
         -------
         data : numpy.ndarray
             Data set.
-        tags : dict
-            Tag information for band
-
         """
+        if nodataval is not None and not isinstance(nodataval, list):
+            nodataval = [nodataval]
+
+        decoder_kwargs = {} if decoder_kwargs is None else decoder_kwargs
+
+        if row is None and col is not None:
+            row = 0
+            n_cols = self.shape[-1]
+        elif row is not None and col is None:
+            col = 0
+            n_rows = self.shape[-2]
+
         if band is None:
-            if sub_rect is None:
+            if row is None and col is None:
                 data = self.src.ReadAsArray()
             else:
-                data = self.src.ReadAsArray(
-                    sub_rect[0], sub_rect[1], sub_rect[2], sub_rect[3])
-        else:
-            gdal_band = self.src.GetRasterBand(int(band))
-            if gdal_band is None:
-                raise IOError("Reading band {:} failed.".format(band))
-            if sub_rect is None:
-                data = gdal_band.ReadAsArray()
-            else:
-                data = gdal_band.ReadAsArray(
-                    sub_rect[0], sub_rect[1], sub_rect[2], sub_rect[3])
-
-        tags = self.read_tags(band)
+                data = self.src.ReadAsArray(col, row, n_cols, n_rows)
 
-        if self.auto_decode:
-            if (tags['scale_factor'] != 1.) or (tags['add_offset'] != 0.):
-                data = data.astype(float)
-                data[data == tags['nodata']] = np.nan
-                data = data*tags['scale_factor'] + tags['add_offset']
-            else:
-                wrn_msg = "Automatic decoding is activated for band '{}', " \
-                          "but attribute 'scale' and 'offset' are missing!".format(band)
-                warnings.warn(wrn_msg)
+            if decoder is not None:
+                if nodataval is None:
+                    nodataval = [-9999]
+                data = decoder(data, nodataval=nodataval[0], **decoder_kwargs)
         else:
-            if encode_func is not None:
-                data = encode_func(data, tags)
+            if not isinstance(band, list):
+                band = [band]
+
+            if nodataval is None:
```

Comment:
isn't it easier to directly set nodataval to -9999

Reply:
From a programming point of view it is easier, but GeoTIFFs do not need to have a no data value